resources/testcases/restricted_block/run.rls

```diff
@@ -1,4 +1,6 @@
+@declare data(any, any).
```

Comment:
Shouldn't the type inference be able to detect this though? I suppose the problem is that "String" is assigned too early as the datatype for the loaded predicate

Reply:
I wonder if the way to fix this is to say something like: 

- Every explicit type declaration triggers type inferences (by that I mean "is a starting point in the dfs-traversel of the predicate-position graph for inferring types)
- Every position which contains an existential variable triggers the type inference
- After this, every unknown type will be replaced by defaults, which will be `any` for idb or the `source.default_type()` for edbs

Reply:
I can give this some thought again.
I definitely wanted to propagate the default source types since otherwise one gets a lot of type clashes as well.
So at the moment the type propagation is done based on the explicit declarations and the data source default types.

Reply:
Yeah you're right; non-any default types would have to be propagated again ... 
Might not be so simple then

Reply:
The existential positions are also propagated now! 
(Going to test with our typical examples/benchmarks.)